### PR TITLE
tc retry mechanism

### PR DESCRIPTION
**Pull request by esquarial**

### CODE REVIEW

File: src/MessageHandler/TestRunMessageHandler.php [hunk 1]
```diff
@@ -25,6 +27,7 @@ class TestRunMessageHandler
     public function __construct(
         private readonly TestRunRepository $testRunRepository,
         private readonly UserRepository $userRepository,
+        private readonly SettingsRepository $settingsRepository,
```
Comment (ppf) [inline src/MessageHandler/TestRunMessageHandler.php:30]: **Blocker:** adding `SettingsRepository` as 3rd ctor arg breaks `tests/Unit/MessageHandler/TestRunMessageHandlerTest.php` — sites at lines 65, 119, 163 still pass the old 9-arg list → `TypeError` at construction.

Fix: add `$this->createStub(SettingsRepository::class)` as 3rd arg in each site.

File: src/MessageHandler/TestRunMessageHandler.php [hunk 2]
```diff
@@ -157,7 +160,22 @@ private function handleExecute(TestRun $run, ?\Closure $receiverLockRefreshCallb
             // Continue to REPORT - service should have marked run as failed
         }
 
-        // Always dispatch REPORT (idempotency in NOTIFY prevents duplicate notifications)
+        // Check for auto-retry before going to REPORT
+        if ($this->shouldAutoRetry($run)) {
+            $retryRun = $this->dispatchAutoRetry($run);
```
Comment (ppf) [inline src/MessageHandler/TestRunMessageHandler.php:165]: **Blocker — double-execution race:** two redelivery windows spawn a second retry chain on the same original run.

1. `dispatchAutoRetry` succeeds (retry `PHASE_PREPARE` queued), parent `PHASE_CLEANUP` dispatch then throws → outer catch at `:115` rethrows → original `PHASE_EXECUTE` redelivers.
2. Visibility-timeout redelivery if heartbeat fails during long `executeRun`.

On redelivery: `shouldAutoRetry` re-evaluates on the original run (`retryAttempt` still `0`) → second retry chain.

Fix via atomic `retry_spawned_at` flag on `retryFailedRun` — see comment there.

File: src/MessageHandler/TestRunMessageHandler.php [hunk 3]
```diff
@@ -277,4 +295,67 @@ private function handleCleanup(TestRun $run): void
...
+        }
+
+        $settings = $this->settingsRepository->getSettings();
+        $maxRetries = $settings->getMaxRetryCount();
+        if ($maxRetries <= 0) {
+            return false;
+        }
+
+        // Check if we've exceeded retry limit
+        if ($run->getRetryAttempt() >= $maxRetries) {
```
Comment (ppf) [inline src/MessageHandler/TestRunMessageHandler.php:348]: **Blocker:** no tests cover `shouldAutoRetry` — the sole defense against infinite retry chains. An off-by-one (`>=` vs `>`) on `getRetryAttempt() >= $maxRetries` would silently flood the queue.

Add tests for each early return + the boundary: `retryAttempt = maxRetries-1` (true), `= maxRetries` (false), `> maxRetries` (false).

File: src/MessageHandler/TestRunMessageHandler.php [hunk 4]
```diff
@@ -277,4 +295,67 @@ private function handleCleanup(TestRun $run): void
             // No next phase - just log and continue
         }
     }
+
+    private function shouldAutoRetry(TestRun $run): bool
```
Comment (ppf) [inline src/MessageHandler/TestRunMessageHandler.php:326]: **Should fix — missing cancellation guard:** `shouldAutoRetry` runs after `executeRun` returns. If a user cancels mid-execution (`executeRun` returns via the cancellation branch at `TestRunnerService.php:284`), failed results from before cancellation are still in DB → returns `true` → spawns an unwanted retry of a cancelled run.

Fix — first early return:
```php
if (TestRun::STATUS_CANCELLED === $run->getStatus()) {
    return false;
}
```

File: src/Service/TestRunnerService.php [hunk 5]
```diff
@@ -20,9 +21,23 @@
  */
 class TestRunnerService
 {
+    private const RETRYABLE_ERROR_PATTERNS = [
```
Comment (ppf) [inline src/Service/TestRunnerService.php:24]: **Blocker — false-positive retries on legitimate assertion failures:** `matchesRetryablePattern` (line 645) runs `stripos` over both `errorMessage` AND the entire output file contents, with bare substrings like `'timed out after'`, `'timed out waiting'`, `'javascript error:'`.

A test asserting `assertStringContainsString('timed out after', ...)` triggers retry on a legitimate assertion failure; the output-file scan also catches unrelated Selenium driver-log timeouts.

Fix:
- Drop the output-file scan.
- Replace bare substrings with exception-class anchors only (`TimeoutException`, `ElementNotInteractableException`, etc.).

File: src/Service/TestRunnerService.php [hunk 6]
```diff
@@ -414,6 +437,67 @@ public function retryRun(TestRun $originalRun, ?User $executedBy = null): TestRu
...
+        return $newRun;
+    }
+
+    /**
+     * Retry only the failed tests from a completed run.
+     *
+     * Analyzes each failed result to determine if it's a retryable infrastructure error.
+     * Returns null if no retryable failures are found.
+     */
+    public function retryFailedRun(TestRun $originalRun, ?User $executedBy = null, int $retryAttempt = 1): ?TestRun
```
Comment (ppf) [inline src/Service/TestRunnerService.php:449]: **Should fix — no DB-level idempotency:** no unique constraint on `(original_run_id, retry_attempt)` and no `retry_spawned_at` flag. Two simultaneous `/retry-failed` clicks both create child runs; same window enables the redelivery race noted on `handleExecute`.

Fix (atomic spawn-flag, mirrors `notification_sent_at` at `TestRunMessageHandler.php:232`):
```php
$affected = $this->connection->executeStatement(
    'UPDATE matre_test_runs SET retry_spawned_at = NOW() WHERE id = ? AND retry_spawned_at IS NULL',
    [$run->getId()]
);
if (0 === $affected) {
    return null;
}
```

File: src/Service/TestRunnerService.php [hunk 7]
```diff
@@ -598,19 +820,88 @@ static function (string $testName) use ($excludedLookup, &$matchedExcluded): boo
...
+            'failedCount' => $failedCount,
+        ]);
+    }
+
+    /**
+     * Execute a single test with inline retry for retryable infrastructure failures.
+     *
+     * @return bool True if the run was cancelled during execution
+     */
+    private function executeSingleTestWithRetry(
```
Comment (ppf) [inline src/Service/TestRunnerService.php:887]: **Should fix — per-attempt logs overwritten:** `executeSingleTest` writes `var/test-output/run-{id}/{safeName}.log` with `w` mode. Each retry attempt overwrites the prior log; persisted `TestResult.outputFilePath` points only to the last attempt — failed-attempt evidence is lost.

Fix: write per-attempt paths (`{safeName}.attempt-{N}.log`).

File: src/Entity/TestRun.php [hunk 8]
```diff
@@ -488,6 +499,76 @@ public function setUpdatedAt(?\DateTimeImmutable $updatedAt): static
...
+     * @param string[] $testIds
+     */
+    public function setRetryTestIdsFromArray(array $testIds): static
+    {
+        $this->retryTestIds = empty($testIds) ? null : json_encode(array_values($testIds));
+
+        return $this;
+    }
+
+    public function isRetry(): bool
```
Comment (ppf) [inline src/Entity/TestRun.php]: **Should fix — dead code:**
- `isRetry()` (this line): no callers in `src/`, `templates/`, `tests/`.
- `getRetryTestIds()` / `setRetryTestIds()` (lines 526, 531): only referenced inside this entity; real callers use `getRetryTestIdsArray()` / `setRetryTestIdsFromArray()`.

Root-pointer expression `$originalRun->getOriginalRun() ?? $originalRun` is also duplicated at `TestRunnerService.php:442` and `:483`.

Fix: delete the unused accessors; add `getRoot(): self` on the entity and replace the duplicated expressions.

File: src/Command/TestRunCommand.php [hunk 9]
```diff
@@ -41,6 +43,7 @@ protected function configure(): void
             ->addOption('filter', 'f', InputOption::VALUE_OPTIONAL, 'Test filter (test name, group, or pattern)')
             ->addOption('suite', 's', InputOption::VALUE_OPTIONAL, 'Test suite name')
             ->addOption('sync', null, InputOption::VALUE_NONE, 'Run synchronously (wait for completion)')
+            ->addOption('retry-failed', null, InputOption::VALUE_NONE, 'Automatically retry infrastructure-failed tests (sync mode only)')
```
Comment (ppf) [inline src/Command/TestRunCommand.php]: **Should fix:** `prepareRun` calls `clearRootLevelArtifacts()` outside the env lock acquired in `executeRun`. If another run is in-flight on the same environment when `--retry-failed` chains a retry, shared artifact dirs get wiped mid-execution.

Fix — document in the option help:
```php
->addOption('retry-failed', null, InputOption::VALUE_NONE, 'Automatically retry infrastructure-failed tests (sync mode only; use only when no other runs are active for this environment)')
```
